### PR TITLE
Fix command line parsing around the -- separator

### PR DESCRIPTION
**Pull request by victorhurdugaci**
Fixes https://github.com/aspnet/dotnet-watch/issues/80, https://github.com/aspnet/dotnet-watch/issues/81

See the two bugs to understand what was broken. There are examples here.

@moozzyk and/or @BrennanConroy can you please review?

Approved for release by @Eilon

### CODE REVIEW

File: test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        [InlineData(new string[] { "--" },
+            new string[0],
+            new string[0])]
+        [InlineData(new string[] { "--appArg1" },
+            new string[0],
+            new string[] { "--appArg1" })]
+        [InlineData(new string[] { "--command", "test" },
+            new string[0],
+            new string[] { "--command", "test" })]
+        [InlineData(new string[] { "--command", "test", "--" },
```
Comment (victorhurdugaci) [inline test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs]: This is the case broken in https://github.com/aspnet/dotnet-watch/issues/80

File: test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            new string[] { "--command", "test" },
+            new string[0])]
+        [InlineData(new string[] { "--command", "test", "--", "--appArg1", "arg1Value" },
```
Comment (victorhurdugaci) [inline test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs]: This is the case broken in https://github.com/aspnet/dotnet-watch/issues/81

File: src/Microsoft.DotNet.Watcher.Tools/Program.cs [hunk 3]
```diff
@@ -101,19 +115,34 @@ private int MainInternal(string[] watchArgs, string[] appArgs, CancellationToken
...
             app.OnExecute(() =>
             {
                 var projectToWatch = Path.Combine(Directory.GetCurrentDirectory(), ProjectModel.Project.FileName);
 
+                var command = commandArg.Value();
+                if (!commandArg.HasValue())
+                {
+                    // The default command is "run". In this case we always assume the arguments are passed to the application being run.
+                    // If you want a different behaviour for run you need to use --command and pass the full arguments
+                    // Run is special because it requires a "--" before the being passed to the app arguments,
```
Comment (moozzyk) [inline src/Microsoft.DotNet.Watcher.Tools/Program.cs]: something is missing in this sentence...

File: test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Xunit;
+
+namespace Microsoft.DotNet.Watcher.Tools.Tests
+{
+    public class ArgumentSeparatorTests
+    {
+        [Theory]
+        [InlineData(new string[0],
```
Comment (moozzyk) [inline test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs]: do you have any test cases for arguments containing whitespaces?

File: src/Microsoft.DotNet.Watcher.Tools/Program.cs [hunk 5]
```diff
@@ -50,31 +49,46 @@ public static int Main(string[] args)
...
+                args[0].Equals("-?", StringComparison.OrdinalIgnoreCase))
             {
                 watchArgs = new string[] { args[0] };
                 appArgs = new string[0];
                 return;
             }
 
-            int argsIndex = -1;
-            watchArgs = args.TakeWhile((arg, idx) =>
+            int separatorIndex = -1;
```
Comment (moozzyk) [inline src/Microsoft.DotNet.Watcher.Tools/Program.cs]: nit: var? (same below)

**Comment by moozzyk** [comment]
minor comments. :ship::it:

File: test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs (hunk 4 above)
Comment (victorhurdugaci) [inline test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs]: Nope...

**Comment by victorhurdugaci** [comment]
Updated

File: src/Microsoft.DotNet.Watcher.Core/DotNetWatcher.cs [hunk 6]
```diff
@@ -58,29 +58,24 @@ public class DotNetWatcher
...
-            else
-            {
-                dotnetArguments = new string[] { command };
-            }
+            var fullDotnetArgs = new string[dotnetArguments.Length + 1];
+            fullDotnetArgs[0] = command;
 
-            dotnetArguments = dotnetArguments
-                .Select(arg =>
+            for (int i = 0; i < dotnetArguments.Length; i++)
```
Comment (muratg) [inline src/Microsoft.DotNet.Watcher.Core/DotNetWatcher.cs]: nit: var

File: test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs (hunk 4 above)
Comment (muratg) [inline test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs]: Should we add this?

Comment (victorhurdugaci) [inline test/Microsoft.DotNet.Watcher.Tools.Tests/ArgumentSeparatorTests.cs]: We cannot test that in a unit test. The quotation for arguments happens in a different part of the code

**Comment by victorhurdugaci** [comment]
Updated again

**Comment by muratg** [comment]
:shipit:

**Pull request merged**